### PR TITLE
Auto shift: support repeats and early registration

### PR DESCRIPTION
**Pull request by p00ya**
## Description

Instead of waiting for the next record to tap the autoshiftable key, register a matrix_scan function that will check the timeout and register the key immediately if the timeout period has elapsed.  This means the user gets immediate feedback about their press and doesn't have to guess whether it's been long enough.

Additionally, don't unregister the keypress immediately.  This means that auto-shifted keys can be held down.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* #7048 

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

I've tested with `AUTO_SHIFT_ENABLE`, `NO_AUTO_SHIFT_ALPHA` on an Ergodox EZ, with taps, holds, and tap-then-holds, and rollovers to autoshiftable and non-autoshiftable keys.  There are a lot of edge cases.

### CODE REVIEW

**Comment by IsaacElenbaas** [comment]
That was fast. Lots of free time in quarantine? :smile: 

> Additionally, don't unregister the keypress immediately. This means that auto-shifted keys can be held down.

I'd argue that the tap-then-hold behavior should also apply to shifted keys (or a flag to change whether it happens on its own). Also, `TAPPING_FORCE_HOLD` should disable the shifted key's autorepeat as well (and probably renamed even without that), as having autorepeat disabled for half of your typing keys seems dumber than all of them.

With this being nearly a rewrite, would you be alright with adding something similar to #8127 to merge in one go? (or is that a bad idea anyway? Kinda new to GitHub) I could code it, just want to try to get custom shifts in as nothing's happened to that for four months. Figuring out autorepeat with shifted keycodes may be tricky.

**Comment by p00ya** [comment]
> That was fast. Lots of free time in quarantine? 😄
> 
> > Additionally, don't unregister the keypress immediately. This means that auto-shifted keys can be held down.
> 
> I'd argue that the tap-then-hold behavior should also apply to shifted keys (or a flag to change whether it happens on its own). Also, `TAPPING_FORCE_HOLD` should disable the shifted key's autorepeat as well (and probably renamed even without that), as having autorepeat disabled for half of your typing keys seems dumber than all of them.

Here's a way of thinking of the functionality I implemented: your auto-shift key is physically stuck to to a `MT(MOD_LSFT, KC_NONE)` key.  Thinking of it this way, I believe `TAPPING_FORCE_HOLD` is doing the expected things (subject to `IGNORE_MOD_TAP_INTERRUPT` and such), i.e. it's just changing whether the MT gets held (presses shift) after tap-then-hold.

> 
> With this being nearly a rewrite, would you be alright with adding something similar to #8127 to merge in one go? (or is that a bad idea anyway? Kinda new to GitHub) I could code it, just want to try to get custom shifts in as nothing's happened to that for four months. Figuring out autorepeat with shifted keycodes may be tricky.

I will punt on it for now...  It took me a few tries (and force pushes) to get an intuitive shift behaviour that I was happy with, which mostly means that the keyboard is faithfully representing its physical state with the exception of the shift key.  Custom shift actions seem much more complicated to think about in the edge cases I've been testing (if I hold "1" then press and release "2", should the keyboard continue to hold the custom shifted "1" or switch it to normal "1"?  What if I only tap "2", effectively requiring shift to be released - will the keyboard be pretending that "1" was released even when it wasn't?).  

I think your feature is cool but probably belongs in tap dance rather than auto-shift.

**Comment by IsaacElenbaas** [comment]
I understand the comparison with the key pressing a MT and itself, but don't see that in the code or a reason to stick to it. I'm tired, might understand in the morning. Either way, I feel a way to disable keyrepeat is necessary.
> I will punt on it for now... It took me a few tries (and force pushes) to get an intuitive shift behaviour that I was happy with, which mostly means that the keyboard is faithfully representing its physical state with the exception of the shift key. Custom shift actions seem much more complicated to think about in the edge cases I've been testing (if I hold "1" then press and release "2", should the keyboard continue to hold the custom shifted "1" or switch it to normal "1"? What if I only tap "2", effectively requiring shift to be released - will the keyboard be pretending that "1" was released even when it wasn't?).

See [here](https://github.com/qmk/qmk_firmware/issues/4072#issuecomment-432431482). AutoShift keys are normally evaluated on release, are they not? You are simply making the timeout an alternate trigger to handle it. Evaluating on 'release' means we have more info about what the user is trying to do. In the case you mention, shift (but not 1) should be released and 2 tapped, then shift restored. In normal typing, this produces the exact effect as with normal keys - hold a, then press b. a will stop being repeated and there will be one b at the end. In a game, it would also likely produce the correct effect, as 1 is not a modifier (and everyone using AS has a gaming layer right now anyway, as there is no keyrepeat). Should 2 be held, it is the thing that would repeat in typing, and so the shift state should belong to it. Both of these fire when the second key is released or hits time, and so updating the shift states and current key is simple as we either swap shift, press, and restore shift without changing anything or (if hit time) swap all AS variables as if the first key did not exist.

Sorry for the wall of text. As I said above, I can code it, and if you don't want it in this pull I can always do it later, but I feel this pull and custom AutoShifts would make the feature truly complete.

**Comment by p00ya** [comment]
> I understand the comparison with the key pressing a MT and itself, but don't see that in the code or a reason to stick to it. I'm tired, might understand in the morning. Either way, I feel a way to disable keyrepeat is necessary.
> 
> > I will punt on it for now... It took me a few tries (and force pushes) to get an intuitive shift behaviour that I was happy with, which mostly means that the keyboard is faithfully representing its physical state with the exception of the shift key. Custom shift actions seem much more complicated to think about in the edge cases I've been testing (if I hold "1" then press and release "2", should the keyboard continue to hold the custom shifted "1" or switch it to normal "1"? What if I only tap "2", effectively requiring shift to be released - will the keyboard be pretending that "1" was released even when it wasn't?).
> 
> See [here](https://github.com/qmk/qmk_firmware/issues/4072#issuecomment-432431482). AutoShift keys are normally evaluated on release, are they not? You are simply making the timeout an alternate trigger to handle it. Evaluating on 'release' means we have more info about what the user is trying to do. In the case you mention, shift (but not 1) should be released and 2 tapped, then shift restored. In normal typing, this produces the exact effect as with normal keys - hold a, then press b. a will stop being repeated and there will be one b at the end. In a game, it would also likely produce the correct effect, as 1 is not a modifier (and everyone using AS has a gaming layer right now anyway, as there is no keyrepeat). Should 2 be held, it is the thing that would repeat in typing, and so the shift state should belong to it. Both of these fire when the second key is released or hits time, and so updating the shift states and current key is simple as we either swap shift, press, and restore shift without changing anything or (if hit time) swap all AS variables as if the first key did not exist.

Cool, that should be consistent with what I've implemented (except maybe the "restore shift" part - this PR won't remember how long you held "1", and you can potentially be still holding it down without it ever being auto-shifted, e.g. by rolling from "1" to "1"+"2" quickly, which I find is common for sequences like "st" on Dvorak).

The main philosophical issue I have with the custom auto shifts (should we call these "custom hold keys" or something instead, since they're not actually sending shifts?) is that if you press a bunch of them together, you have to know when you pressed each one to figure out what the keyboard might actually be sending (if your fingers are on "1" and "2", but autoshift_custom_shifts(KC_1) == KC_7, what's QMK reporting as being held down?).  In the simpler pure auto-shift case, you only need to look at where your fingers are and when you pressed the *last* key.  Pragmatically (for typing and such), this probably doesn't matter, but it seemed like a nice theoretical goal when I was staring at a keyboard viewer when testing.

I'll need to have more of a think about whether that's compatible with preventing repeats.  The intuitive behaviour for me is that if you hold a key down QMK should faithfully report that key as being held down to the OS.

> Sorry for the wall of text. As I said above, I can code it, and if you don't want it in this pull I can always do it later, but I feel this pull and custom AutoShifts would make the feature truly complete.

Please do go ahead and code it - it doesn't really matter what I think, plenty of QMK features don't follow my "stay true to the physical keyboard" philosophy!  Whether I merge it into this PR will depend on what the QMK folks say about both - in the worst case we can have dependent PRs and rebase.

I also suspect that this PR needs to be cleaned up a little bit, particularly to support custom hold actions (there are a bunch of assumptions about what's going on with the shift key which will break when the hold actions aren't actually sending shifts).  Regardless of the custom hold actions, cleaning this PR up is a good thing, so if you organize your commits in a way that makes it easy to merge cleanups/refactors that don't actually change the behaviour, I'd be interested in merging those.  You can send PRs to my branch (though note I'm using a rebase workflow).

**Comment by IsaacElenbaas** [comment]
> if you press a bunch of them together, you have to know when you pressed each one to figure out what the keyboard might actually be sending (if your fingers are on "1" and "2", but autoshift_custom_shifts(KC_1) == KC_7, what's QMK reporting as being held down?). In the simpler pure auto-shift case, you only need to look at where your fingers are and when you pressed the last key. Pragmatically (for typing and such), this probably doesn't matter, but it seemed like a nice theoretical goal when I was staring at a keyboard viewer when testing.

You seem to view AutoShift as a way to automate the shift layer, and above describe issues with it being "half-active". I want it to _remove_ the shift layer, and simply be a fast way to make every key have a different hold action (otherwise you have to use an advanced tapdance function [and, even if there was a better way, either lots of defines or XX(KC_X, LSFT(KC_X)) for every key]). It's just that the obvious second value for typing keys are their capital versions, and so the old shift layer's value for that key is the default. Describing new ones is necessary, on a small board exclamation point as period's second value makes way more sense than carat.

Also, I couldn't get the tap hold to repeat the unshifted character even with TAPPING_TERM over 1s. Don't see why it isn't working for me, the code flow looks correct.

File: quantum/process_keycode/process_auto_shift.c
Comment (IsaacElenbaas) [inline quantum/process_keycode/process_auto_shift.c:167]: This needs right shift added if my cleanup isn't merged.
```suggestion
        if (keycode == KC_LSFT || keycode == KC_RSFT) {
```

**Comment by IsaacElenbaas** [comment]
https://github.com/p00ya/qmk_firmware/pull/1

File: quantum/process_keycode/process_auto_shift.c
Comment (p00ya) [reply quantum/process_keycode/process_auto_shift.c:167]: The auto-shifting here only ever holds down LSFT.  If the user presses RSFT, it shouldn't interfere with auto-shift and can just be processed normally.

Comment (IsaacElenbaas) [reply quantum/process_keycode/process_auto_shift.c:167]: I thought that holding LSFT, holding a key, and releasing would unshift while RSFT would not. I don't know why, but I just tested and that's not the case. Tapping either while repeating shifted also unshifts. You're right, not an issue.

**Comment by p00ya** [comment]
> You seem to view AutoShift as a way to automate the shift layer, and elsewhere also describe issues with it being "half-active". I want it to remove the shift layer, and simply be a fast way to make every key have a different hold action (otherwise you have to use an advanced tapdance function [and, even if there was a better way, either lots of defines or XX(KC_X, LSFT(KC_X)) for every key]).

Well, yes, I think automating the shift layer is strongly implied by the name "AutoShift"! ;)

Have you considered rebranding "custom auto shift" as "custom hold actions" or similar?

> [p00ya#1](https://github.com/p00ya/qmk_firmware/pull/1)
> If that is merged, adding custom shifts should be a trivial task. The only things to figure out is release behavior (release both? Record what was pressed for that custom key?) and getting the keycodes to it.

Per my philosophy above, my personal preference would be per-key release, which probably means more state to track.

> 
> For the latter there are a few options.
> 
> * Number one, which won't work, is `AS(KC_A, LSFT(KC_B))`. Though this would probably be best, there isn't room to store two keys with mods in a `uint16` (same issue as [#2055 (comment)](https://github.com/qmk/qmk_firmware/pull/2055#issuecomment-376758268)). Even rearranging to only pass the key's shift state wouldn't work, and that wouldn't support foreign layouts which need more to send certain keys.

You can hack around that the same way custom macros work: allocate a new keycode, and have a table that expands that to whatever you need (just a struct with 2 keycodes?).

One nice feature of this is that you can have two keys with the same keycode for the "tap" action but different keycodes for the "hold" action.

> * Number two is [TapDance's workaround.](https://beta.docs.qmk.fm/using-qmk/software-features/feature_tap_dance#simple-example-id-simple-example) I haven't looked into it and don't know how hard it would be to make, but this could give AutoShift more information on what's actually happening vs just what to hit.
> * Number three is just a user-defined function, as in the original pull (#8127). On the odd return values, I've realized I'm dumb and it can just return the actual thing to press, so this isn't ugly and has the same functionality as number two (one could define a macro key and no space issue).
> 
> Thoughts?

I think a user-defined function (perhaps with a predefined one respecting AUTO_SHIFT_ALPHAS etc. for compatibility) would be clean and pragmatic.

Though I do still think the custom hold actions are something that more properly belongs in tap dance, e.g. if it the hold was part of `qk_tap_dance_state_t`.

Probably out of scope but worth mentioning: the really advanced option is to hook right into `action_tapping.c`.  I considered that when rewriting auto-shift.  But that code is even more fragile and less readable than what's in quantum/, which is probably why everyone is content to hook downstream in `process_record`.

**Comment by IsaacElenbaas** [comment]
> Well, yes, I think automating the shift layer is strongly implied by the name "AutoShift"! ;)

> Though I do still think the custom hold actions are something that more properly belongs in tap dance, e.g. if it the hold was part of `qk_tap_dance_state_t`.

I know, it felt wrong as I was writing it (how can you misinterpret 'AutoShift'?) but really, what is AS except a single-pattern from tap dance with defaults? There should be support for the most basic of configuration, especially since AS+TD is a good chunk of your available space and doing more than a few custom through TD pollutes the keymap so recreating AS isn't practical.

> I think a user-defined function (perhaps with a predefined one respecting AUTO_SHIFT_ALPHAS etc. for compatibility) would be clean and pragmatic.

I'm planning on leaving the current `process_auto_shift`, with just a `if custom` at the top so that alphas could be overridden.
Actually, I just had a thought. Functions could be called on press, evaluation (with bool timeout passed or not), *and* release, and they press/release the key and return its keycode (which is recorded and sent to release later). I currently have some hacked-together keys that are auto-shifted, but also act as mod-taps. Things like that would be easy to make, and all three functions would be optional, so if you don't define `eval` and `release`, what was returned by `down` is released for custom keys.

I get that this is going deep into TapDance territory, but that's all AutoShift is. Heck, it probably would have been a good idea to make AutoShift just enable TD and create a bunch of advanced functions for whatever AUTO_SHIFT_XXXX are defined (isn't TD smaller firmware size even before this pull?). I'll code this up tomorrow, if I don't hit any roadblocks I think it could be pretty clean.

**Comment by IsaacElenbaas** [comment]
This should no longer be a breaking change due to to flag changes, aside from the firmware size increase.

An option to get AS to return true for later key processing is having a constant custom keycode and editing the record to be it on any AS key down/up. Not sure where it would change anything that can be used parallel to AS, but could be useful later. That would allow other things to see that they shouldn't handle the key (how would they?) but know that the physical event happened.

**Comment by noroadsleft** [comment]
This needs to be rebased upon `develop` and have the merge conflicts resolved.

**Comment by noroadsleft** [comment]
Sorry; need another rebase/conflict fix here. There was another PR for Auto Shift that got merged.

**Comment by p00ya** [comment]
@noroadsleft ping for merge before you force-push develop again

**Comment by fauxpark** [comment]
Develop and master are now locked. This PR will have to go in next cycle.

**Comment by p00ya** [comment]
Rebased yet-again.

Question for reviewers: this PR introduces a new preprocessor symbol to opt-in to the new behaviour (`AUTO_SHIFT_REPEAT`), and another symbol `AUTO_SHIFT_NO_AUTO_REPEAT` to have tap-then-hold repeating but not hold-only repeating.  If you're comfortable with breaking changes, `AUTO_SHIFT_REPEAT` could be changed to opt-out, since I think repeat is the most natural behaviour.

Another option is to use `TAPPING_FORCE_HOLD` instead of (or as a default value for) `AUTO_SHIFT_REPEAT`.  This would control the behaviour of tapped-then-held auto shift keys: if set it would send the shifted key on the hold instead of holding the unshifted key, much like a MT will send the modifier on the hold.  However, this will get a bit more complicated to support `TAPPING_FORCE_HOLD_PER_KEY`.

**Comment by IsaacElenbaas** [comment]
We discussed `TAPPING_FORCE_HOLD_PER_KEY` a bit in the original custom AS keys pull and decided support should be added later. I'm planning on closing that once this is merged and reopening with my fork from this, which I designed with `TAPPING_FORCE_HOLD_PER_KEY` in mind. It should be very easy to implement.

**Comment by p00ya** [comment]
My main concern is that introducing `AUTO_SHIFT_REPEAT` is introducing additional API surface and complexity, which sounds like it will get deprecated pretty soon anyway.  Rather than introducing this legacy config option now, it seems better to reuse `TAPPING_FORCE_HOLD` and `TAPPING_FORCE_HOLD_PER_KEY` straight away, even if this means changing the semantics for existing users who already have `AUTO_SHIFT_ENABLE` on today.  But I think we need the qmk folks to weigh in on what their policy is for changes in behaviour like this.

**Comment by IsaacElenbaas** [comment]
I just read the docs again and realized I was thinking of something else, `TAPPING_FORCE_HOLD_PER_KEY` could be added before custom shifts. Even if we just changed the define to `TAPPING_FORCE_HOLD`, though, there's little harm in having a period of time where `PER_KEY` doesn't work - I can think of few situations where I would need it here and none would be in existing keymaps.

What will be odd is `AUTO_SHIFT_NO_AUTO_REPEAT`. I think it should exist, but it will be related to `TFH` and only affect AS (as `TFH` is mainly for mod-taps).

**Comment by noroadsleft** [comment]
K, I've been testing this off-and-on since Saturday, and I want to make sure I understand this correctly, as I'm fairly useless at C and reading the actual code isn't helping me. :slightly_frowning_face: 

Changes:

- The key being triggered is automatically sent when shifted, instead of waiting for key release
- `AUTO_SHIFT_REPEAT` repeats the shifted character if the key is not released after `AUTO_SHIFT_TIMEOUT` ms; repeats the normal character if tapped then held
- `AUTO_SHIFT_NO_AUTO_REPEAT` disables the automatic repeating of `AUTO_SHIFT_REPEAT`, but can still be triggered manually by holding the key after the initial tap is registered

Am I missing anything? I've literally never used Auto Shift before.

**Comment by IsaacElenbaas** [comment]
Correct, and with `AUTO_SHIFT_NO_AUTO_REPEAT` the tap hold repeat uses the shift state of the tap (so you can repeat aaaa or AAAA).
> reading the actual code isn't helping me

Probably my fault, lol, sorry. It used to be much easier to wrap your head around but the current approach is needed to make custom shifts possible in the future without a rewrite. Planning on turning https://github.com/p00ya/qmk_firmware/pull/2 into its own pull once this is merged.

**Comment by noroadsleft** [comment]
K, I think I'm hitting some weird behaviour with `AUTO_SHIFT_NO_AUTO_REPEAT` defined and `AUTO_SHIFT_REPEAT` *undefined* – it's extremely difficult to get an unshifted key with these settings. I'm thinking `AUTO_SHIFT_NO_AUTO_REPEAT` isn't intended to be used without `AUTO_SHIFT_REPEAT`, but when I try it, the uppercase/lowercase status of the key that is input seems to be dependent on how long it's been since the last key was entered, as if there's a timer running that's continually toggling the state of Shift.

All other states seem to be working as intended.

**Comment by IsaacElenbaas** [comment]
Hmm, I can't reproduce and everything referencing `AUTO_SHIFT_NO_AUTO_REPEAT` is nested under `AUTO_SHIFT_REPEAT`. Are you sure that was what caused it?

**Comment by noroadsleft** [comment]
Sorry for the slow responses; I've had three different issues with my PC in the last week that have affected my ability to work on QMK. :slightly_frowning_face: 

I've been testing this by creating a branch based upon this PR, and adding a commit which adds a keymap specifically for testing this change. I have a Proton C mounted to a breadboard, with an MX-type switch connected (wires are soldered to the switch legs).

I originally tried only setting `AUTO_SHIFT_TIMEOUT` to 180ms. I just took that setting and changed it to 500ms, which should make it very rare to get a shifted key, but tapping the switch actually gave me shifted keys more often than not. Here's an output string from testing:

```
AaAAAaaAaaAAaAaA
```

This is with me not holding the switch down at all. Typical typing speeds don't activate keys for longer than half a second, so I still think something is up.

This is the diff between this branch and the one I'm testing from:

```diff
diff --git a/keyboards/handwired/onekey/keymaps/nrl/config.h b/keyboards/handwired/onekey/keymaps/nrl/config.h
new file mode 100644
index 0000000000..54f0651f23
--- /dev/null
+++ b/keyboards/handwired/onekey/keymaps/nrl/config.h
@@ -0,0 +1,13 @@
+#pragma once
+
+#undef MATRIX_ROWS
+#undef MATRIX_COLS
+#define MATRIX_ROWS 1
+#define MATRIX_COLS 7
+
+#undef MATRIX_COL_PINS
+#undef MATRIX_ROW_PINS
+#define MATRIX_COL_PINS { A1, A0, B8, B13, B14, B15, B9 }
+#define MATRIX_ROW_PINS { A2 }
+
+#define AUTO_SHIFT_TIMEOUT 500
diff --git a/keyboards/handwired/onekey/keymaps/nrl/keymap.c b/keyboards/handwired/onekey/keymaps/nrl/keymap.c
new file mode 100644
index 0000000000..ad6c2c5423
--- /dev/null
+++ b/keyboards/handwired/onekey/keymaps/nrl/keymap.c
@@ -0,0 +1,7 @@
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+    LAYOUT_ortho_1x1(
+        KC_A, KC_B, KC_C, KC_D, KC_E, KC_F, KC_G
+    )
+};
diff --git a/keyboards/handwired/onekey/keymaps/nrl/rules.mk b/keyboards/handwired/onekey/keymaps/nrl/rules.mk
new file mode 100644
index 0000000000..c9383ab8db
--- /dev/null
+++ b/keyboards/handwired/onekey/keymaps/nrl/rules.mk
@@ -0,0 +1 @@
+AUTO_SHIFT_ENABLE = yes
diff --git a/keyboards/handwired/onekey/onekey.h b/keyboards/handwired/onekey/onekey.h
index 72cab45be5..bfa3cc2cfa 100644
--- a/keyboards/handwired/onekey/onekey.h
+++ b/keyboards/handwired/onekey/onekey.h
@@ -19,7 +19,7 @@
 #include "quantum.h"
 
 #define LAYOUT_ortho_1x1( \
-    k00 \
+    k00, k01, k02, k03, k04, k05, k06 \
 ) { \
-    { k00 } \
+    { k00, k01, k02, k03, k04, k05, k06 } \
 }
diff --git a/keyboards/handwired/onekey/proton_c/config.h b/keyboards/handwired/onekey/proton_c/config.h
index 3ba4ba6499..e4ebb6310b 100644
--- a/keyboards/handwired/onekey/proton_c/config.h
+++ b/keyboards/handwired/onekey/proton_c/config.h
@@ -27,6 +27,6 @@
 #define BACKLIGHT_PWM_CHANNEL   3
 #define BACKLIGHT_PAL_MODE      2
 
-#define RGB_DI_PIN A0
+//#define RGB_DI_PIN A0
 
-#define ADC_PIN A0
+//#define ADC_PIN A0
```

**Comment by IsaacElenbaas** [comment]
Does any combination of `#define`s work properly for you? Above you blamed `AUTO_SHIFT_NO_AUTO_REPEAT` without `AUTO_SHIFT_REPEAT` but neither are in that diff. If `#define`s are the issue, it shouldn't be hard to find, as they really don't affect much.
It may be worth it to `send_string(elapsed)` [here](https://github.com/qmk/qmk_firmware/blob/70dc57ed7df57ed42627d97148fefdf94eb1d03d/quantum/process_keycode/process_auto_shift.c#L99) and test with both short and large delays between taps, (if you can `send_string` an int). @p00ya, have you tried testing recently? Everything is working fine on a barebones keymap on my Planck rev6 but it could be a board thing. Would test on my Corne but I've lost the heavily modified `process_auto_shift.c` used for its keymap.

**Comment by noroadsleft** [comment]
> Does any combination of `#define`s work properly for you? Above you blamed `AUTO_SHIFT_NO_AUTO_REPEAT` without `AUTO_SHIFT_REPEAT` but neither are in that diff. If `#define`s are the issue, it shouldn't be hard to find, as they really don't affect much.

Ah, I forgot to mention that I was testing with each combination of `AUTO_SHIFT_NO_AUTO_REPEAT` and `AUTO_SHIFT_REPEAT` – both off, both on, `AUTO_SHIFT_REPEAT` without `AUTO_SHIFT_NO_AUTO_REPEAT`, and `AUTO_SHIFT_NO_AUTO_REPEAT` without `AUTO_SHIFT_REPEAT`. I then reset to just default settings with a 500ms `AUTO_SHIFT_TIMEOUT` (no `AUTO_SHIFT_NO_AUTO_REPEAT` or `AUTO_SHIFT_REPEAT` defined), but I'm still getting shifted keys on taps.

I'll try the `send_string()`.

**Comment by noroadsleft** [comment]
Well I wasn't able to get the `send_string()` debug to work – again, I am *useless* at C – but even trying to tap the key once every second still gives me shifted characters, with nothing defined other than `AUTO_SHIFT_TIMEOUT 500`.

It seems like the first time Auto Shift fires, a timer starts running that toggles the shift state on an interval, and it doesn't matter how long I have the key pressed the next time I use it.

If I can get my phone sorted out I can see about taking a video tomorrow.

**Comment by p00ya** [comment]
Did some testing myself as a sanity check, using my Ergodox EZ, the PR branch and ergodox_ez:default (with the local edits to rules.mk and config.h below).  Everything seems to work for me.

I wonder if there's something weird about the matrix scan rate or the clock on the Proton C test keyboard.  I might try asking folks on discord about the way we're using `TIMER_DIFF_16` and `record->event.time`, since those things are keyboard specific.

@noroadsleft as a sanity check, can you confirm auto-shift works from the develop branch *without* this PR for you?

### AUTO_SHIFT_ENABLE = yes without additional defines

Tap: `a`
Hold:  `A`
Tap-and-hold: `aA`
Shift tap: `A`
Shift hold: `A`
Shift tap and hold: `AA`
Hold then tap other: `Ao`
Hold then hold other: `AO`

### with just AUTO_SHIFT_REPEAT

Tap: `a`
Hold:  `AAA`
Tap-and-hold: `aaa`
Shift tap: `A`
Shift hold: `AAAA`
Shift tap and hold: `AAAAA`
Hold then tap other: `AAAAo`
Hold then hold other: `AAAAAOOOO`

### with AUTO_SHIFT_REPEAT and AUTO_SHIFT_NO_AUTO_REPEAT

Tap: `a`
Hold:  `A`
Tap-and-hold: `aaa`
Shift tap: `A`
Shift hold: `A`
Shift tap and hold: `AAAAA`
Hold then tap other: `Ao`
Hold then hold other: `AO`

### with just AUTO_SHIFT_NO_AUTO_REPEAT

Tap: `a`
Hold:  `A`
Tap-and-hold: `aA`
Shift tap: `A`
Shift hold: `A`
Shift tap and hold: `AA`
Hold then tap other: `Ao`
Hold then hold other: `AO`

**Comment by noroadsleft** [comment]
> @noroadsleft as a sanity check, can you confirm auto-shift works from the develop branch _without_ this PR for you?

Good call.

---

## Running on `develop` branch, *without* this PR

### AUTO_SHIFT_ENABLE = yes without additional defines (2020-11-22 00:35 -08:00)

- Tap: `x`
- Hold: `X`
- Tap-and-hold: `xX`
- Shift tap: `X`
- Shift hold: `X`
- Shift tap and hold: `XX`
- Hold then tap other: ?? (don't know yet)
- Hold then hold other: ??

---

*to be continued...*

**Comment by noroadsleft** [comment]
## Running on `p00ya:pull-auto-shift` with `#define AUTO_SHIFT_TIMEOUT 500` on a Proton C

### AUTO_SHIFT_ENABLE = yes without additional defines (2020-11-22 12:16 -08:00)

* Tap: `a`
* Hold: `A`
* Tap-and-hold: `aA aA aA AA aA aA aA aA AA AA`
* Shift tap: `A`
* Shift hold: `A`
* Shift tap and hold: `AA`
* Hold then tap other: `AX AX AX AX Ax AX AX Ax Ax Ax AX AX Ax`
* Hold then hold other: `AXAX`

Repeating certain Auto Shift actions seems to give me the strange behaviour.  When testing *Hold then tap other*, sometimes I get a shifted input as soon as I press the key down, without having to wait for `AUTO_SHIFT_TIMEOUT`. Additionally, sometimes I get a shifted key on the *tap other*, as if the shift hasn't been disabled yet. *Tap and hold* also exhibits similar behaviour.

What's the timing for when the Shift is disabled after an Auto Shift key fires? I've also now tested this on a Pro Micro, and it's doing the same thing.

**Comment by IsaacElenbaas** [comment]
Shift is disabled immediately, and a report manually sent (on ANY key event after press), so it's not that shift isn't being released.
It seems as though `autoshift_time` is occasionally not being set (or set to `0`) on key down, looking at your `tap-and-hold` and `hold then tap other` tests - note you never have accidentally *unshifted* keys. We store `record->event.time` and compare with `TIMER_DIFF_16` (passing current time from release event (or `timer_read` if matrix) too), while the original AS stores using `timer_read` and `timer_elapsed`. Perhaps `event.time` is inconsistent/record is freed? I don't see a reason for the original to poll for the current time on both press and release when it has the time right there otherwise.

Could you test with every `TIMER_DIFF_16(...)` replaced with `timer_elapsed(autoshift_time)` and every `record->event.time` with `timer_read()`?

**Comment by noroadsleft** [comment]
> Could you test with every `TIMER_DIFF_16(...)` replaced with `timer_elapsed(autoshift_time)` and every `record->event.time` with `timer_read()`?

**Success!**

## Same settings as my last comment, tested on Proton C and Pro Micro

* Tap-and-hold: `aA aA aA aA aA aA aA aA aA aA`
* Hold then tap other: `Ax Ax Ax Ax Ax Ax Ax Ax Ax Ax`

All other sequences are the same as before.

**Comment by p00ya** [comment]
Thanks for testing.  I'll update the PR.

That certainly is weird behaviour, though, and seems to me like a bug with the keyboard implementation if `record->event.time` is incorrectly set.

**Comment by p00ya** [comment]
Please test again.  I've kept the `TIMER_DIFF_16` but am now performing a `timer_read()` from `process_auto_shift`.

**Comment by noroadsleft** [comment]
> Please test again. I've kept the `TIMER_DIFF_16` but am now performing a `timer_read()` from `process_auto_shift`.

Everything's still good, same results I had before.

**Review by noroadsleft — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**
